### PR TITLE
Force type hints

### PR DESCRIPTION
**Pull request by finswimmer**
This PR add missing type hints to both package code and test code. Furthermore `flake-annotations` is added as a flake8 plugin to enforce type hints in the future.

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -1,7 +1,11 @@
 [flake8]
```
Comment (neersighted) [inline .flake8:1]: Thsi whole file is very cargo-culted. Can we clean it up while we're in here?

File: .flake8 [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
 max-line-length = 88
-ignore = E501, E203, W503
-per-file-ignores = __init__.py:F401
+ignore = E501, E203, W503, ANN101, ANN102
+mypy-init-return = True
```
Comment (neersighted) [inline .flake8:4]: Is this a mypy config setting? If so, please put it in pyproject.toml (see #199).

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -9,11 +9,13 @@ repos:
     rev: 3.9.2
     hooks:
       - id: flake8
+        additional_dependencies: [flake8-annotations]
```
Comment (neersighted) [inline .pre-commit-config.yaml:12]: We shouldn't need additional exclusions here if the `.flake8` ones are good.

File: poetry/core/version/pep440/version.py [hunk 4]
```diff
@@ -46,7 +47,21 @@ def __post_init__(self):
 
         object.__setattr__(self, "_compare_key", self._make_compare_key())
 
-    def _make_compare_key(self):
+    def _make_compare_key(
```
Comment (neersighted) [inline poetry/core/version/pep440/version.py:50]: Wow, this is a complicated type. Is there any way to clean this up with NewType, a refactor with an object, or similar?

**Review by neersighted — CHANGES_REQUESTED** [review]

File: tests/conftest.py [hunk 5]
```diff
@@ -10,7 +11,12 @@
 from tests.testutils import tempfile
 
 
-def pytest_addoption(parser):
+if TYPE_CHECKING:
```
Comment (danieleades) [inline tests/conftest.py:14]: you shouldn't need the `if TYPE_CHECKING` pattern here- that's really only for preventing circular imports. You can just import these directly, and use them below without quoting the types

File: tests/conftest.py [hunk 6]
```diff
@@ -20,15 +26,15 @@ def pytest_addoption(parser):
     )
 
 
-def pytest_configure(config):
+def pytest_configure(config: "Config") -> None:
```
Comment (danieleades) [inline tests/conftest.py:29]: you can use `Config` unquoted if you remove the `if TYPE_CHECKING` pattern above

File: .flake8 (hunk 2 above)
Comment (finswimmer) [reply .flake8:4]: Nope, it's a `flake8-annotation` setting. See here: https://pypi.org/project/flake8-annotations/

> Allow omission of a return type hint for `__init__` if at least one argument is annotated. See mypy's documentation for additional details.

File: tests/json/test_poetry_schema.py [hunk 7]
```diff
@@ -4,7 +4,7 @@
 
 
 @pytest.fixture
-def base_object():
+def base_object() -> dict:
```
Comment (danieleades) [inline tests/json/test_poetry_schema.py:7]: this syntax is python 3.9+. less than that, the correct return type would be `Dict`, rather than `dict`.

I guess we can get away with it, since `mypy` is being run with a recent interpreter, and the type is ignored during other testing, but whichever way we go, it should be consistent throughout the codebase

File: tests/masonry/builders/test_builder.py [hunk 8]
```diff
@@ -10,7 +13,11 @@
 from poetry.core.utils._compat import PY37
 
 
-def test_builder_find_excluded_files(mocker):
+if TYPE_CHECKING:
```
Comment (danieleades) [inline tests/masonry/builders/test_builder.py:16]: is there a circular import if you do not guard the import of `MockerFixture`?

File: .pre-commit-config.yaml (hunk 3 above)
Comment (finswimmer) [reply .pre-commit-config.yaml:12]: I thought that too. But it looks like when running `flake8` via `pre-commit` the exclusions in `.flake8` are ignored :man_shrugging: Would be happy to see only one single source of truth.

**Review by danieleades — CHANGES_REQUESTED** [review]
mypy itself can be configured to return errors on missing annotations, i believe. is the 'flake8' plugin somehow preferable?

- `if TYPE_CHECKING` pattern doesn't appear to be required throughout
- inconsistent use of `dict`/`Dict`

File: poetry/core/version/pep440/version.py (hunk 4 above)
Comment (finswimmer) [reply poetry/core/version/pep440/version.py:50]: I was happy, that PyCharm was doing this for me :laughing: IMO refactoring should be done in another PR, because the purpose of this PR is to enforce type hints in the future.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (neersighted) [reply .pre-commit-config.yaml:12]: That was not my experience with the main repo. I suspect what's happening is partial vs. full paths being passed. Try fragments like e.g. `*/tests/fixtures/*/setup.py`. Some experimentation is likely needed.

File: poetry/core/version/pep440/version.py (hunk 4 above)
Comment (neersighted) [reply poetry/core/version/pep440/version.py:50]: That is a brutal type signature :)

File: tests/json/test_poetry_schema.py (hunk 7 above)
Comment (danieleades) [reply tests/json/test_poetry_schema.py:7]: while we're at it, these `dict` annotations are equivalent to `Dict[Any, Any]`- we should be able to do better than that

File: tests/testutils.py [hunk 9]
```diff
@@ -70,8 +70,8 @@ def subprocess_run(*args, **kwargs):  # type: (str, Any) -> subprocess.Completed
 
 
 def validate_wheel_contents(
-    name, version, path, files=None
-):  # type: (str, str, str, Optional[List[str]]) -> None
+    name: str, version: str, path: str, files: Optional[List[str]] = None
+) -> None:
     dist_info = "{}-{}.dist-info".format(name, version)
     files = files or []
```
Comment (danieleades) [inline tests/testutils.py:76]: does the following work?

```suggestion
    name: str, version: str, path: str, files: [List[str] = []
) -> None:
    dist_info = "{}-{}.dist-info".format(name, version)
```

Comment (finswimmer) [reply tests/testutils.py:76]: [Python Mutable Defaults Are The Source of All Evil](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/) :smiley:

Comment (danieleades) [reply tests/testutils.py:76]: > [Python Mutable Defaults Are The Source of All Evil](https://florimond.dev/en/posts/2018/08/python-mutable-defaults-are-the-source-of-all-evil/) smiley

ah. yep.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (finswimmer) [reply .pre-commit-config.yaml:12]: Ok, I've got it now. I removed the exclusion definitions completely from from yml file and make the `.flake8` to the singe source if truth.

File: tests/json/test_poetry_schema.py (hunk 7 above)
Comment (finswimmer) [reply tests/json/test_poetry_schema.py]: As you find out `dict` is valid as well, but don't let us define how keys and values look like in python<3.9. I agree it's in inconsistent use if we use `Dict` in other places, so I changed it.

> while we're at it, these dict annotations are equivalent to Dict[Any, Any]- we should be able to do better than that

The only way to do it better here is to use `Dict[str, Any]` (which doesn't add much benefit) or use `TypedDict`. But this is only available for python >=3.8 and would require to add `typing-extensions` as dependency for python <3.8. Not sure if we want to introduce another dependency.

File: poetry/core/version/pep440/parser.py [hunk 10]
```diff
@@ -61,7 +61,9 @@ def _get_local(cls, match: Optional[Match[AnyStr]]) -> Optional[LocalSegmentType
         )
 
     @classmethod
-    def parse(cls, value: str, version_class: Optional[Type["PEP440Version"]] = None):
+    def parse(
+        cls, value: str, version_class: Optional[Type["PEP440Version"]] = None
+    ) -> "PEP440Version":
```
Comment (danieleades) [inline poetry/core/version/pep440/parser.py:66]: if you remove the `if TYPE_CHECKING` guard, this can be written as-

```suggestion
    def parse(
        cls, value: str, version_class: Optional[PEP440Version] = None
    ) -> PEP440Version:
```

File: tests/conftest.py (hunk 5 above)
Comment (finswimmer) [reply tests/conftest.py:14]: That's a convention I'm used to. If you need a module just for type hints, import it only in the `if TYPE_CHECKING` block. 

It's easier to teach to other people ("Why are you sometime put import in that block and sometimes not?") and the overall chance that you have a circular import if you need this import just for type hint is quite big.

TL;DR: I prefer to leave it as it is.

File: poetry/core/version/pep440/version.py (hunk 4 above)
Comment (finswimmer) [reply poetry/core/version/pep440/version.py:50]: I've tried to understand the code and IMO the signature can be written more easily. See code change.

File: tests/conftest.py (hunk 5 above)
Comment (danieleades) [reply tests/conftest.py:14]: fair enough. there's some interesting background on it here - https://github.com/larryhastings/co_annotations/issues/1

in general, it seems that `if TYPE_CHECKING` is used iff

- you would otherwise have a circular import
- the import is 'expensive'

the Benevolent Dictator for Life seems to regard it is a painfully necessary hack.

that said, it's also clear from the thread that it's far from settled. And using it *everywhere* instead of the small number of cases where you need it (do you definitely need it anywhere in this codebase?) is at least consistent.

**Comment by danieleades** [comment]
might be a good idea to merge #199 before this is merged, in order to catch any type errors that could be interested by this PR

**Comment by neersighted** [comment]
re: `is TYPE_CHECKING`, check out https://github.com/asottile/flake8-typing-imports

Looks like a useful set of lints we could apply here.

**Review by neersighted — APPROVED** [review]
